### DISCUSSION TITLE
How can/should this be used with nextjs?

### DISCUSSION THREAD

**Discussion by mmkal** [TanStack/router#346] [Q&A] [answered]
Hi! I've idly browsed through the docs for react-location and thought this sounds great, but we wouldn't consider abandoning nextjs completely, given the many benefits we get from it (and the many we expect from next 13 and beyond). Is there a recommended way to use react-location in nextjs projects, or even a hint of how to start thinking about things? Or are they fundamentally incompatible? Would be really interesting to see a nextjs project in the examples folder.

I saw https://github.com/TanStack/router/discussions/44 and it was resolved with a link to a comparison page, but that only seems to compare with react-router.

I also found https://github.com/TanStack/router/discussions/95 which asks "can this be used with nextjs" and the answer was "yes". Which is great, but would be great to know more about how, or if there are any pitfalls to look out for.

**Comment by tannerlinsley** [comment, marked as answer]
Right now, TanStack Router is not designed to be used with full-stack frameworks and is best suited for client-side only apps. In the future, we will be addressing SSR and framework integration. With regards to those issues, you *can* use client-side routers inside of Next and other full-stack frameworks, but you are essentially opting out of the SSR patterns and benefits you get from their routers.

**Reply by puterhimself** [reply]
is this still true with the latest release?

**Reply by stychu** [reply]
Would like to know too

**Reply by gruckion** [reply]
@tannerlinsley still true?

**Comment by FlorentinoG9** [comment]
take a look at this in the mean time 😀 probably it can help you

https://next-typesafe-url.dev/

**Comment by enricoros** [comment]
Would love to hear an update on this - i.e. the recommended policy for Migrating from Next.JS, or maybe wait until **Start** is out, or is it not an option? Have a few-pages Next.js app (pages router) that I'd love to convert.

**Reply by enricoros** [reply]
Note: I'll be happy to Fully migrate away from NextJS and SSR as the vendor lock-in is not sustainable and the limitations of SSR don't work for my App. Hence, I'd be looking for a medium-pain migration path from *client-side only* (and local-first <3) routing.

**Comment by RafaRochaS91** [comment]
@tannerlinsley we can however still use this if we treat for example a page of NextJS as its own self contained app with CSR and Client Side Routing. 
This would be a potential use case to have examples in place for since I've seen quite a few companies using NextJS as more of an orchestration than an actual full stack framework (perhaps this was wrong from the get go so please elucidate me).

**Reply by dvidsilva** [reply]
Astro officially supports this. Should be the same to try it out in Next right? 

https://www.reddit.com/r/reactjs/comments/1jr8nn7/made_a_poc_for_building_spa_with_astro_and/

**Reply by RafaRochaS91** [reply]
more or less, the way astro uses islands for client side boundary and interactivity makes this integration way smoother since there's no RSC mess, but you're right - its possible to do a similar approach. What i've ended up doing (unfortunately I have to stick to NextJS since its an existing application):

1. create a nextjs app route
2. define the route definition for that prefix within the app router folder
3. just import and mount client based components from lib in the sub route definitions.

This kinda works but it means for potential other "anchor" points of tanstack this would need to have multiple entry points - root routes per NextJS server side route.

I will create an example that I can publicly share and link it here soon.
